File: python/310.py
"""Project Euler Problem 310 - Nim Square (Python 3.12)

This module computes the number of losing positions in the Nim Square game for
triples (a, b, c) with 0 <= a <= b <= c <= MAX.

Nim Square rules:
- Normal play, 3 heaps.
- A move consists of removing a perfect-square number of stones from a heap.

We use standard Sprague-Grundy theory:
- Precompute Grundy numbers for single-heap sizes up to MAX.
- A triple (a, b, c) is losing iff g(a) ^ g(b) ^ g(c) == 0.

The original Ruby file contained several iterative "improved" attempts and
verbose logging; this Python version exposes a clean, tested API while keeping
an executable "main" routine.
"""

from __future__ import annotations

from bisect import bisect_left
from dataclasses import dataclass
from functools import lru_cache
from math import isqrt
from typing import Dict, Iterable, List, Sequence, Tuple


@dataclass(frozen=True)
class NimSquareConfig:
    """Configuration for Nim Square computations.

    Attributes:
        max_heap: Inclusive upper bound on heap size (c in 0 <= a <= b <= c).
    """

    max_heap: int = 100_000

    def __post_init__(self) -> None:  # type: ignore[override]
        if self.max_heap < 0:
            msg = "max_heap must be non-negative"
            raise ValueError(msg)

    @property
    def squares(self) -> List[int]:
        """List of allowed move sizes (perfect squares up to max_heap)."""

        limit = isqrt(self.max_heap)
        return [i * i for i in range(1, limit + 1)]


def compute_grundy_table(config: NimSquareConfig) -> List[int]:
    """Compute Grundy numbers g[n] for 0 <= n <= max_heap.

    Uses a bottom-up dynamic program specialized for moves of square sizes.
    The original Ruby used recursive memoization with Set; we use an
    iterative approach, which is both clearer and faster in Python.
    """

    max_n = config.max_heap
    squares = config.squares

    grundy = [0] * (max_n + 1)

    # Temporary boolean array reused for mex computation.
    # Size is conservative: Grundy values stay reasonably small in practice.
    # For safety, sized relative to number of squares (upper bound on degree).
    max_candidates = len(squares) + 2
    seen = [False] * max_candidates

    for n in range(1, max_n + 1):
        # Mark reachable Grundy values from n.
        # Only squares <= n are valid moves.
        for sq in squares:
            if sq > n:
                break
            g = grundy[n - sq]
            if g < max_candidates:
                seen[g] = True

        # mex: smallest non-negative integer not in seen.
        mex = 0
        while mex < max_candidates and seen[mex]:
            mex += 1
        grundy[n] = mex

        # Reset seen for values we may have touched.
        for sq in squares:
            if sq > n:
                break
            g = grundy[n - sq]
            if g < max_candidates:
                seen[g] = False

    return grundy


def verify_small_case() -> bool:
    """Verify correctness for MAX=29 where the answer is known (1160).

    Returns True if the precomputed Grundy table reproduces the known count.
    """

    config = NimSquareConfig(max_heap=29)
    grundy = compute_grundy_table(config)

    count = 0
    for a in range(0, 30):
        ga = grundy[a]
        for b in range(a, 30):
            gb = grundy[b]
            for c in range(b, 30):
                gc = grundy[c]
                if (ga ^ gb ^ gc) == 0:
                    count += 1

    return count == 1160


def build_positions_by_grundy(grundy: Sequence[int]) -> Tuple[List[int], List[List[int]]]:
    """Build frequency and positions lists for each Grundy value.

    Returns a tuple (freq, positions):
    - freq[g] = how many heap sizes n have Grundy value g
    - positions[g] = sorted list of such n
    """

    max_g = max(grundy)
    freq = [0] * (max_g + 1)
    positions: List[List[int]] = [[] for _ in range(max_g + 1)]

    for n, g in enumerate(grundy):
        freq[g] += 1
        positions[g].append(n)

    return freq, positions


def _count_ordered_triples_for_values(
    pos_a: Sequence[int], pos_b: Sequence[int], pos_c: Sequence[int]
) -> int:
    """Count triples (a, b, c) with a <= b <= c.

    - a ranges over pos_a
    - b ranges over pos_b
    - c ranges over pos_c

    All sequences must be sorted. This function is performance-critical and
    intentionally compact.
    """

    if not pos_a or not pos_b or not pos_c:
        return 0

    count = 0

    for b in pos_b:
        # a choices: indices where a <= b
        # bisect_left returns first index with value > b, so that index
        # equals the count of entries <= b.
        num_a = bisect_left(pos_a, b + 1)

        # c choices: indices where c >= b
        idx_c = bisect_left(pos_c, b)
        num_c = len(pos_c) - idx_c

        if num_a and num_c:
            count += num_a * num_c

    return count


def count_losing_positions(config: NimSquareConfig) -> int:
    """Count losing positions (a, b, c) for Nim Square under given config.

    A position is losing for the next player iff g[a] ^ g[b] ^ g[c] == 0.
    The algorithm:
    - Compute Grundy table g[n].
    - Group heap sizes by Grundy value.
    - For each pair (ga, gb), derive gc = ga ^ gb and count triples using
      precomputed position lists.

    Complexity is roughly O(MAX + G^3), where G is the number of distinct
    Grundy values, which is small in practice for this game.
    """

    grundy = compute_grundy_table(config)
    freq, positions = build_positions_by_grundy(grundy)
    max_g = len(freq) - 1

    total = 0

    for ga in range(max_g + 1):
        if not freq[ga]:
            continue
        pos_a = positions[ga]

        for gb in range(max_g + 1):
            if not freq[gb]:
                continue
            pos_b = positions[gb]

            gc = ga ^ gb
            if gc > max_g or not freq[gc]:
                continue
            pos_c = positions[gc]

            total += _count_ordered_triples_for_values(pos_a, pos_b, pos_c)

    return total


def main(max_heap: int = 100_000) -> None:
    """Entry point for CLI-style execution.

    - Verifies correctness on the small MAX=29 case.
    - Computes and prints the losing-position count for given max_heap.
    """

    config = NimSquareConfig(max_heap=max_heap)

    if not verify_small_case():
        msg = "Verification for MAX=29 failed; aborting computation."
        raise RuntimeError(msg)

    result = count_losing_positions(config)
    print(result)


if __name__ == "__main__":  # pragma: no cover - CLI behavior
    import argparse

    parser = argparse.ArgumentParser(
        description=(
            "Compute the number of losing positions in Nim Square for "
            "0 <= a <= b <= c <= MAX."
        ),
    )
    parser.add_argument(
        "MAX",
        type=int,
        nargs="?",
        default=100_000,
        help="Maximum heap size (default: 100000)",
    )
    args = parser.parse_args()

    main(args.MAX)

File: c/310.c
/*
 * Project Euler Problem 310: Nim Square
 *
 * Count losing positions (a,b,c) with 0<=a<=b<=c<=100000
 * where g[a]^g[b]^g[c]==0, g being Grundy numbers for Nim Square
 * (moves are perfect squares).
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define MAX_HEAP 100000

static int grundy[MAX_HEAP + 1];

/* Precompute squares */
static int squares[320]; /* sqrt(100000) ~ 316 */
static int nsquares;

static void compute_grundy(void) {
    int sq_limit = (int)sqrt((double)MAX_HEAP) + 1;
    nsquares = 0;
    for (int i = 1; i <= sq_limit; i++)
        squares[nsquares++] = i * i;

    /* Temporary seen array for mex. Grundy values stay small. */
    int max_cand = nsquares + 2;
    char *seen = calloc(max_cand, 1);

    grundy[0] = 0;
    for (int n = 1; n <= MAX_HEAP; n++) {
        for (int i = 0; i < nsquares && squares[i] <= n; i++) {
            int g = grundy[n - squares[i]];
            if (g < max_cand) seen[g] = 1;
        }
        int mex = 0;
        while (mex < max_cand && seen[mex]) mex++;
        grundy[n] = mex;

        /* Reset */
        for (int i = 0; i < nsquares && squares[i] <= n; i++) {
            int g = grundy[n - squares[i]];
            if (g < max_cand) seen[g] = 0;
        }
    }
    free(seen);
}

int main(void) {
    compute_grundy();

    /* Find max Grundy value */
    int max_g = 0;
    for (int i = 0; i <= MAX_HEAP; i++)
        if (grundy[i] > max_g) max_g = grundy[i];

    /* Build position lists by Grundy value */
    int *freq = calloc(max_g + 1, sizeof(int));
    for (int i = 0; i <= MAX_HEAP; i++)
        freq[grundy[i]]++;

    /* Build sorted position arrays */
    int **positions = calloc(max_g + 1, sizeof(int *));
    int *pos_idx = calloc(max_g + 1, sizeof(int));
    for (int g = 0; g <= max_g; g++) {
        if (freq[g] > 0)
            positions[g] = malloc(freq[g] * sizeof(int));
    }
    for (int i = 0; i <= MAX_HEAP; i++) {
        int g = grundy[i];
        positions[g][pos_idx[g]++] = i;
    }

    /* Count triples (a,b,c) with a<=b<=c and g[a]^g[b]^g[c]==0 */
    long long total = 0;

    for (int ga = 0; ga <= max_g; ga++) {
        if (!freq[ga]) continue;
        int *pos_a = positions[ga];
        int na = freq[ga];

        for (int gb = 0; gb <= max_g; gb++) {
            if (!freq[gb]) continue;
            int gc = ga ^ gb;
            if (gc > max_g || !freq[gc]) continue;

            int *pos_b = positions[gb];
            int nb = freq[gb];
            int *pos_c = positions[gc];
            int nc = freq[gc];

            /* Count triples with a in pos_a, b in pos_b, c in pos_c, a<=b<=c */
            /* For each b, count a's <= b in pos_a, count c's >= b in pos_c */
            int ia = 0, ic_start = 0;
            for (int ib = 0; ib < nb; ib++) {
                int b = pos_b[ib];
                /* Count a's <= b: binary search or linear scan */
                while (ia < na && pos_a[ia] <= b) ia++;
                int num_a = ia; /* pos_a[0..ia-1] are <= b */

                /* Count c's >= b: find first index in pos_c >= b */
                while (ic_start < nc && pos_c[ic_start] < b) ic_start++;
                int num_c = nc - ic_start;

                if (num_a > 0 && num_c > 0)
                    total += (long long)num_a * num_c;
            }
        }
    }

    printf("%lld\n", total);

    for (int g = 0; g <= max_g; g++)
        free(positions[g]);
    free(positions);
    free(pos_idx);
    free(freq);

    return 0;
}
